### DISCUSSION TITLE
Hidden Attribute

### DISCUSSION THREAD

**Discussion by dbpolito** [spatie/laravel-data#472] [Ideas]
I do have some DTOs that i have some internal properties that i never want to expose, i'm currently using `exceptProperties` within the class with does the job, but i think a better DX would be having a `Hidden` attribute to annotate which ones i don't want to expose.

Are you guys open for a PR for that?

**Comment by rubenvanassche** [comment]
If it's well tested and documented I'm always open for PR's 😄

**Comment by plakhin** [comment]
@dbpolito could you share the implementation please? I'm facing the issue mentioned in #353

**Comment by dbpolito** [comment]
@plakhin for not i'm just adding this:

```php
    public function exceptProperties(): array
    {
        return [
            'attribute' => true,
        ];
    }
```

but i just created a PR for a Hidden attribute, let see if we get this merged: https://github.com/spatie/laravel-data/pull/505

**Comment by plakhin** [comment]
> ```php
>    public function exceptProperties(): array
>    {
>        return [
>            'attribute' => true,
>        ];
>    }

@dbpolito, this way properties are also not being saved to the DB in case an eloquent model casts an attribute to Data Object.


And since you're modifying `resolvePayload()` of `DataTransformer` it looks like Hidden attribute will also be ("hidden") not saved to the database.

**Comment by dbpolito** [comment]
hey @plakhin , thanks for the feedback... i have pushed some changes that should fix it... i wasn't aware of this feature 😬

**Comment by rubenvanassche** [comment]
As said in the PR, the hide attribute should work like except otherwise we create inconsistencies and nobody will understand how any of this works.

An option to transform data without running these excepts, onlys and hides is a better solution in this case since it creates clarity. Maybe I'll add this to v4, but PR's are always welcome.

**Comment by plakhin** [comment]
>  otherwise we create inconsistencies 

True

@rubenvanassche 
As mentioned in #353 I'm facing an issue that I wasn't able to solve in any way including your proposal and I thought #505 could help. However, you're right that keeping features consistent is more important.

**Reply by rubenvanassche** [reply]
You can always write your own Eloquent cast which hides some properties

**Reply by plakhin** [reply]
And use it instead of a class that extends `Data` class? 
Sorry, just can't understand.

**Reply by rubenvanassche** [reply]
You can extend the DataEloquentCast so that it hides certain properties, and then use such class by overwriting `castUsing` if I'm not mistaken within you own Data class

**Reply by plakhin** [reply]
Got it. Thanks!
